### PR TITLE
Add support for offline planning of variable tensors

### PR DESCRIPTION
**Pull request by mansnils**
* New assumption that offline planner (in this case Vela) supports variable tensors
 * Move variable tensor allocation in the online planner to later in the memory plannign phase so that we have the opportunity to get the offfsets from the offline planner for these as well.

Co-authored-by: Fredrik Svedberg fredrik.svedberg@arm.com

BUG=Offline planner does not support variable tensors.

### CODE REVIEW

File: tensorflow/lite/micro/micro_allocation_info.cc [hunk 1]
```diff
@@ -204,6 +204,10 @@ TfLiteStatus AllocationInfoBuilder::InitializeAllocationInfo(
           (current->bytes != 0);
       if (offline_offsets) {
         current->offline_offset = offline_offsets[i];
+        if (subgraph->tensors()->Get(i)->is_variable() &&
```
Comment (advaitjain) [inline tensorflow/lite/micro/micro_allocation_info.cc:207]: can we add a comment explaining this? it seems subtle.

File: tensorflow/lite/micro/micro_allocator.cc [hunk 2]
```diff
@@ -819,6 +814,16 @@ TfLiteStatus MicroAllocator::CommitStaticMemoryPlan(
   const int32_t* offline_planner_offsets = nullptr;
   TF_LITE_ENSURE_STATUS(
       builder.GetOfflinePlannedOffsets(&offline_planner_offsets));
+
+  // Allocate buffers for variable tensors.
```
Comment (advaitjain) [inline tensorflow/lite/micro/micro_allocator.cc:818]: let's add a comment about why the variable allocation needs to be here. (i.e. something about its interaction with the offline planner)

**Review by advaitjain — CHANGES_REQUESTED** [review]

File: tensorflow/lite/micro/micro_allocation_info.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocation_info.cc:210]: We need to set needs_allocating here only for offline planned variable tensors and not for all tensors. Added a comment.

File: tensorflow/lite/micro/micro_allocator.h [hunk 3]
```diff
@@ -248,8 +248,9 @@ class MicroAllocator {
   virtual TfLiteStatus AllocateTfLiteEvalTensors(
       const Model* model, SubgraphAllocations* subgraph_allocations);
   // Allocates persistent tensor buffers for variable tensors in the subgraph.
-  virtual TfLiteStatus AllocateVariables(const SubGraph* subgraph,
-                                         TfLiteEvalTensor* eval_tensors);
+  virtual TfLiteStatus AllocateVariables(
+      const SubGraph* subgraph, TfLiteEvalTensor* eval_tensors,
+      const int32_t* offline_planner_offsets = nullptr);
```
Comment (advaitjain) [inline tensorflow/lite/micro/micro_allocator.h:253]: is this default argument needed? doesn't seem like it.

File: tensorflow/lite/micro/micro_allocator.h [hunk 4]
```diff
@@ -248,8 +248,9 @@ class MicroAllocator {
   virtual TfLiteStatus AllocateTfLiteEvalTensors(
       const Model* model, SubgraphAllocations* subgraph_allocations);
   // Allocates persistent tensor buffers for variable tensors in the subgraph.
-  virtual TfLiteStatus AllocateVariables(const SubGraph* subgraph,
-                                         TfLiteEvalTensor* eval_tensors);
+  virtual TfLiteStatus AllocateVariables(
```
Comment (advaitjain) [inline tensorflow/lite/micro/micro_allocator.h:251]: can you add a comment about what the offline_planner_offsets parameter is for?

File: tensorflow/lite/micro/recording_micro_allocator.h [hunk 5]
```diff
@@ -77,8 +77,9 @@ class RecordingMicroAllocator : public MicroAllocator {
       const Model* model, SubgraphAllocations* subgraph_allocations) override;
   TfLiteStatus AllocateTfLiteEvalTensors(
       const Model* model, SubgraphAllocations* subgraph_allocations) override;
-  TfLiteStatus AllocateVariables(const SubGraph* subgraph,
-                                 TfLiteEvalTensor* eval_tensors) override;
+  TfLiteStatus AllocateVariables(
+      const SubGraph* subgraph, TfLiteEvalTensor* eval_tensors,
+      const int32_t* offline_planner_offsets = nullptr) override;
```
Comment (advaitjain) [inline tensorflow/lite/micro/recording_micro_allocator.h:82]: same comment here. let's remove the default nullptr parameter and instead change RecordingMicroInterpreter to use the updated API.

**Review by advaitjain — CHANGES_REQUESTED** [review]
sorry missed a few small things in my last review around not having default arguments.

File: tensorflow/lite/micro/micro_allocator.h (hunk 3 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator.h]: no removing it

File: tensorflow/lite/micro/micro_allocator.h (hunk 4 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator.h:254]: done

File: tensorflow/lite/micro/recording_micro_allocator.h (hunk 5 above)
Comment (mansnils) [reply tensorflow/lite/micro/recording_micro_allocator.h]: Done

**Comment by mansnils** [comment]
All comments resolved and ready for review.

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

**Pull request merged**